### DISCUSSION TITLE
Openssl libcrypto high CPU usage

### DISCUSSION THREAD

**Discussion by shettyh** [Kong/kong#10544] [Help]
Using a plugin to parse and encode JWT token using `openssl.digest`, `openssl.pkey`, during the perf observed that ~50% of CPU is used by `libcrypto`

<img width="564" alt="Screenshot 2023-03-20 at 12 56 06 PM" src="https://user-images.githubusercontent.com/13759916/227105495-00e6fcb0-74cd-4150-a6bc-7c92f91bceba.png">

is this expected ? Anyway to optimise this ?

###More info

- Kong version : 2.8.3
- OpenSSL version: Bundled with Kong ( 1.1.1o )
- OS: Alpine Linux

**Comment by ghost** [comment]
Interesting question, can you provide a minimal reproducible example?

**Reply by shettyh** [reply]
@sabertobihwy thanks for the reply. Please find the sample plugin code below

imports ( tried `resty.openssl` also )
```
local openssl_digest = require "openssl.digest"
local openssl_pkey = require "openssl.pkey"
```

Inside handler

```
local signature = openssl_pkey.new(key):sign(openssl_digest.new("sha256"):update(signing_input))
```

Please let me know if you need more details

**Reply by ghost** [reply]
what is your complete `perf` command that you are using?

**Reply by shettyh** [reply]
Just using default [perf](https://phoenixnap.com/kb/linux-perf) command and executing `perf top`

**Reply by ghost** [reply]
I did a preliminary verification and did not find this phenomenon. I set `nginx_worker_processes = 1` in `kong.conf,` which will only start one worker process. Then I used `sudo perf top -p 235619` to monitor Kong, where` 235619` is the only worker PID.

**Reply by ghost** [reply]
You can use the latest version of Kong to verify this issue on Ubuntu and similar images. I'm always unsure about the libcrypto in Alpine images.

**Reply by ghost** [reply]
And we can more precisely specify perf monitoring for the dso `libcrypto.so.1.1`.

1. find which `libcrypto.so.1.1` that Kong used:

```shell
sudo lsof -p 235619 | grep libcrypto.so.1.1
nginx   235619  sabertobihwy  mem       REG                8,2  7027136  5637100 /home/sabertobihwy/workspace/kong-dev/openssl/lib/libcrypto.so.1.1
``` 

`235619` is the worker PID as above.

2. watch by perf

```shell
 sudo perf top -p 235619 --dsos /home/sabertobihwy/workspace/kong-dev/openssl/lib/libcrypto.so.1.1
```

In my test scenario, I continuously send requests with a valid JWT to Kong for verification, and the CPU usage of `libcrypto.so.1.1` remains low.

**Reply by shettyh** [reply]
@sabertobihwy Thanks for the detailed response. I wont be able to try with latest kong as it requires other changes.

Tried suggested steps in 2.8.3 alpine image, it is using the `libcrypto` that is bundled with kong ( `/usr/local/kong/lib/libcrypto.so.1.1` ).  Will try with ubuntu image also once to verify the same

Bdw, my current setup is
- 1 Kong node ( c5.2x large, 8 core CPU, 16GB RAM)
- Nginx workers -> auto
- Traffic RPS:  Gradually increasing upto 500 RPS over 10 mins

**Reply by ghost** [reply]
> Will try with ubuntu image also once to verify the same

Waiting for your verification result.

Overall, I think `libcrypto` shouldn't consume too much CPU, and it's more likely related to specific platforms.

**Reply by shettyh** [reply]
@sabertobihwy verified in Ubuntu image ( Kong 2.8.3 ), seeing same pattern as Alpine image. Able to process only around 500 RPS in 8 core machine. If i disable the this plugin, same setup is able to handle north of 2K RPS.

To give you more context, this plugin is responsible for generating the JWT token and pass this as header to upstream. As i mentioned before, `perf` is showing `libcrypto` as bottleneck 

Any pointers on what could be wrong and how to debug further ?

**Reply by ghost** [reply]
I need you to provide a minimal, reproducible example as detailed as possible so that I can reproduce the issue locally.

**Reply by fffonion** [reply]
Hi @sabertobihwy , the top three calls are RSA computation using x86 assembly, which itself is expected and heavy workload and is implemented by openssl (libcrypto.so). So your obversation is expected.
If you are concerned about the performance impact when using jwt token, considering constraint client to only send tokens other than `RS256`, `RS384` and `RS512` ones. Or consider using hardware accesslaration cards. You can use `openssl speed` command to check the performance of your platform.

**Reply by shettyh** [reply]
@fffonion thanks for the response. Need some clarity on few things

>considering constraint client to only send tokens other than RS256, RS384 and RS512 ones

Using other algorithms for signing ? You mean like `ecdsa` or similar ?

- Is there any difference between between `openssl` and `resty.openssl` libraries for lua ?
- Any other alternative to improve the performance ?

**Reply by fffonion** [reply]
@shettyh Yes, ECDSA is generally considered faster than RSA under same security level. 

> Is there any difference between between openssl and resty.openssl libraries for lua ?
Using `openssl` (luaossl) under LuaJIT (which Kong uses) is not JIT-able, `resty.openssl` most likely would (I haven't verified every code path though).

> Any other alternative to improve the performance ?
I would suggest to start with the `openssl speed rsa` benchmark, note `RS256` for example is RSA + SHA256, although sha256 is usually hardward accelerated but it wouldn't hurt to benchmark it as well. With that would you give you an idea on how many rounds of operations your platform would get. Double check if CPU of doing cryptography is the bottheneck here, if so, I would check `/proc/cpuinfo` to see if it's missing modern CPu accerlation extensions; then adding hardware is the next option : )

**Reply by fffonion** [reply]
I did a benchmark on my aws 8core dev machine with `openssl speed -multi 8 rsa` for referenece:
```
version: 3.0.2
built on: Mon Feb  6 17:57:17 2023 UTC
options: bn(64,64)
compiler: gcc -fPIC -pthread -m64 -Wa,--noexecstack -Wall -Wa,--noexecstack -g -O2 -ffile-prefix-map=/build/openssl-hnAO60/openssl-3.0.2=. -flto=auto -ffat-lto-objects -flto=auto -ffat-lto-objects -fstack-protector-strong -Wformat -Werror=format-security -DOPENSSL_TLS_SECURITY_LEVEL=2 -DOPENSSL_USE_NODELETE -DL_ENDIAN -DOPENSSL_PIC -DOPENSSL_BUILDING_OPENSSL -DNDEBUG -Wdate-time -D_FORTIFY_SOURCE=2
CPUINFO: OPENSSL_ia32cap=0xfed83203078bffff:0x209c01a9
                  sign    verify    sign/s verify/s
rsa  512 bits 0.000012s 0.000001s  85057.1 1319750.3
rsa 1024 bits 0.000037s 0.000002s  26929.6 456593.6
rsa 2048 bits 0.000252s 0.000008s   3967.5 131980.0
rsa 3072 bits 0.000801s 0.000016s   1247.7  61425.8
rsa 4096 bits 0.001852s 0.000029s    539.9  35007.3
rsa 7680 bits 0.016681s 0.000098s     59.9  10239.5
rsa 15360 bits 0.090580s 0.000383s     11.0   2608.4
```

**Reply by shettyh** [reply]
@fffonion thanks for the response. I will try with `ECDSA` once and post the result.

In the above performance result you are using all 8 cores, is the behaviour same when used within kong/nginx also ? or it always uses single core ?

**Reply by fffonion** [reply]
@shettyh Kong is able to use multicore under default configurations yes. (See https://docs.konghq.com/gateway/latest/reference/configuration/#nginx_worker_processes)

**Reply by shettyh** [reply]
@fffonion i understand that, but i am asking in the context where a plugin is using `openssl` library for signing and that will be running inside one of the nginx worker. In this case also the `openssl` invocation uses multiple cores ?

**Reply by fffonion** [reply]
@shettyh No, a single core is used for a single request. Even openssl cannot use multicore to do a single sign/verify with multicore, so does the `openssl speed` benchmark above.

**Reply by shettyh** [reply]
Ok great, thanks for clarifying
